File: backend/src/main/java/project/backend/global/security/dto/CustomOAuth2User.java
package project.backend.global.security.dto;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import lombok.AllArgsConstructor;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.oauth2.core.user.OAuth2User;

@RequiredArgsConstructor
public class CustomOAuth2User implements OAuth2User {

	private final OAuth2User oAuth2User;
	private final String email;
	private final String githubAccess;

	@Override
	public Map<String, Object> getAttributes() {
		Map<String, Object> extended = new HashMap<>(oAuth2User.getAttributes());
		extended.put("githubAccess", githubAccess);
		extended.put("email", email);
		return extended;
	}

	@Override
	public Collection<? extends GrantedAuthority> getAuthorities() {
		return oAuth2User.getAuthorities();
	}

	@Override
	public String getName() {
		return oAuth2User.getName();
	}
}

File: backend/src/main/java/project/backend/auth/dto/CustomOAuth2User.kt
package project.backend.auth.dto

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.oauth2.core.user.OAuth2User

data class CustomOAuth2User(
    val oAuth2User: OAuth2User,
    val githubAccess: String
) : OAuth2User {

    override fun getAttributes(): Map<String, Any> =
        oAuth2User.attributes.toMutableMap().apply {
            put("githubAccess", githubAccess)
        }

    override fun getAuthorities(): Collection<GrantedAuthority> =
        oAuth2User.authorities

    override fun getName(): String = oAuth2User.name
}